lib/bolt/applicator.rb

```diff
@@ -157,6 +156,10 @@ def apply(args, apply_body, scope)
       plan_vars = scope.to_hash
       %w[trusted server_facts facts].each { |k| plan_vars.delete(k) }
 
+      # Get variables explicitly set to undef from plan scope
+      undef_symbols = scope.effective_symtable(false).instance_variable_get(:@symbols).select { |_sym, val| val.nil? }
```

Comment:
This doesn't seem to work if the `apply()` is inside another scope, since it only uses the most local symtable.

```puppet
plan test(
  TargetSpec $nodes,
  Optional[String] $myvar = undef
) {
  1.each |$x| {
    $nodes.apply_prep
    apply($nodes) {
      notice("My var is ${myvar}")
    }
  }
}
```

Reply:
ahh, do you think `effective_symtables` is the way to go? I think I can collect `@symbols` from by following the `@parent` attributes for that. Though I wonder what the impact of modifying the `to_hash` method to not exclude `undef` variable references would be.